### PR TITLE
Add options to rewrite namespace delimiter for pulsar sql.

### LINKED ISSUE #4732: Support for other schema name separators in pulsar SQL
Pulsar sql table schema using the same naming policy as pulsar namespace, but "/" will cause some problem with presto connection string.

Here is a snapshot of superset use pulsar sql as datasource
image

>Note: Can be replace with presto://172.30.91.40:8081/pulsar?schema=public/default

Get table schema error in superset use pulsar sql as datasource
image

So i would like to suggest pulsar SQL table schema can support other named separators like "-" or "|"  will bring convenience to users.

I'm interested to add this feature.

### PR DESCRIPTION
**Pull request by codelipenghui**
### Motivation

Fix #4732 

### Modifications

Add options to rewrite the namespace delimiter, disable by default

Enable rewrite namespace delimiter can work well with superset:
superset

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: conf/presto/catalog/pulsar.properties [hunk 1]
```diff
@@ -30,7 +30,10 @@ pulsar.target-num-splits=2
 # max message queue size
 pulsar.max-split-message-queue-size=10000
 # max entry queue size
-pulsar.max-split-entry-queue-size = 1000
+pulsar.max-split-entry-queue-size=1000
+# Rewrite namespace delimiter
+pulsar.namespace-delimiter-rewrite-enable=false
+pulsar.rewrite-namespace-delimiter=__
```
Comment (jerrypeng) [inline conf/presto/catalog/pulsar.properties]: Lets leave the default "/"

File: pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/PulsarConnectorConfig.java [hunk 2]
```diff
@@ -40,8 +40,12 @@
     private int maxSplitEntryQueueSize = 1000;
     private int maxMessageSize = Commands.DEFAULT_MAX_MESSAGE_SIZE;
     private String statsProvider = NullStatsProvider.class.getName();
+
     private Map<String, String> statsProviderConfigs = new HashMap<>();
 
+    private boolean namespaceDelimiterRewriteEnable = false;
+    private String rewriteNamespaceDelimiter = "__";
```
Comment (jerrypeng) [inline pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/PulsarConnectorConfig.java]: same here

**Comment by jerrypeng** [comment]
@codelipenghui I have encountered this before its because superset doesn't escape these characters correctly.  Its more of a superset problem than pulsar problem.

**Comment by sijie** [comment]
@codelipenghui keep the default separator as "/" for backward compatibility and allow overwriting with other separator.

@jerrypeng we need this overwritten feature. so +1 for having this feature. but agreed with keeping the default as "/".

**Comment by codelipenghui** [comment]
@jerrypeng @sijie
I have change the default delimiter to "/",  PTAL.

**Comment by codelipenghui** [comment]
run cpp tests

**Comment by jerrypeng** [comment]
@sijie @codelipenghui sure we can have this feature.  Lets add some comments in pulsar.properties for the config because it could be very dangerous and potentially render all queries to fail since tenant and namespace  names can contain all sort of characters e.g. "_" or "-".  If the delimiter is set to a symbol that already exists in the tenant name or namespace name.   

Perhaps we can also we can also have logic to check the delimiter character to make sure it cannot be contained in a valid tenant or namespace name

**Comment by jerrypeng** [comment]
here are the valid characters a tenant and namespace name can be:

https://github.com/apache/pulsar/blob/master/pulsar-common/src/main/java/org/apache/pulsar/common/naming/NamedEntity.java

**Comment by codelipenghui** [comment]
run cpp tests
run java8 tests
run Integration tests

**Comment by codelipenghui** [comment]
run java8 tests
run Integration tests

**Comment by sijie** [comment]
run cpp tests

File: pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/PulsarConnectorConfig.java [hunk 3]
```diff
@@ -191,6 +195,26 @@ public PulsarConnectorConfig setOffloaderProperties(String offloaderProperties)
         return this;
     }
 
+    public String getRewriteNamespaceDelimiter() {
+        return rewriteNamespaceDelimiter;
+    }
+
+    @Config("pulsar.rewrite-namespace-delimiter")
+    public PulsarConnectorConfig setRewriteNamespaceDelimiter(String rewriteNamespaceDelimiter) {
+        this.rewriteNamespaceDelimiter = rewriteNamespaceDelimiter;
```
Comment (jerrypeng) [inline pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/PulsarConnectorConfig.java]: @codelipenghui can you write a simple check to make sure the delimiter is not an allowed character for tenant and namespace names.

File: pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/PulsarConnectorFactory.java [hunk 4]
```diff
@@ -49,7 +49,7 @@ public ConnectorHandleResolver getHandleResolver() {
     @Override
     public Connector create(String connectorId, Map<String, String> config, ConnectorContext context) {
         requireNonNull(config, "requiredConfig is null");
-        log.debug("Creating Pulsar connector with configs: %s", config);
+        log.info("Creating Pulsar connector with configs: %s", config);
```
Comment (jerrypeng) [inline pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/PulsarConnectorFactory.java]: Do we need to change this to info level?  Presto will automatically log the connector configs

**Comment by sijie** [comment]
run cpp tests

**Comment by jiazhai** [comment]
run cpp tests
run integration tests

**Comment by codelipenghui** [comment]
@jerrypeng I have addressed your comments, PTAL.

**Comment by codelipenghui** [comment]
run cpp tests

**Comment by codelipenghui** [comment]
run cpp tests

**Comment by codelipenghui** [comment]
run java8 tests

**Comment by codelipenghui** [comment]
run java8 tests

**Comment by sijie** [comment]
run java8 tests

**Comment by sijie** [comment]
run java8 tests

**Comment by sijie** [comment]
run java8 tests

**Comment by sijie** [comment]
@codelipenghui 

It seems that the java8 test failures are related to this pull request. PTAL

```
Test Result (15 failures / +15)
org.apache.pulsar.sql.presto.TestPulsarMetadata.testGetColumnHandles
org.apache.pulsar.sql.presto.TestPulsarMetadata.testGetTableHandle
org.apache.pulsar.sql.presto.TestPulsarMetadata.testGetTableMetadata
org.apache.pulsar.sql.presto.TestPulsarMetadata.testGetTableMetadataTableBlankSchema
org.apache.pulsar.sql.presto.TestPulsarMetadata.testGetTableMetadataTableInvalidSchema
org.apache.pulsar.sql.presto.TestPulsarMetadata.testGetTableMetadataTableNoSchema
org.apache.pulsar.sql.presto.TestPulsarMetadata.testGetTableMetadataWrongSchema
org.apache.pulsar.sql.presto.TestPulsarMetadata.testGetTableMetadataWrongTable
org.apache.pulsar.sql.presto.TestPulsarMetadata.testListSchemaNames
org.apache.pulsar.sql.presto.TestPulsarMetadata.testListTable
```

**Comment by codelipenghui** [comment]
run cpp tests

**Comment by codelipenghui** [comment]
run cpp tests
run java8 tests

**Comment by codelipenghui** [comment]
run java8 tests

**Comment by jerrypeng** [comment]
@codelipenghui @sijie in regards to this merged PR, what happens if the topic name has a "-"? In SQL you will have to delimit that as well with quotes e.g.

select * from public."public/default"."test-topic"

Even if you change the namespace delimiter from "/" to something to prevent needing to use quotes to delimit, you will still run into the issue with the topic name.

**Comment by jiazhai** [comment]
mark this as 2.4.1 for conflict #4779

**Pull request merged**